File: lib/py/telepact/internal/types/VStruct.py
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from ..validation.ValidateContext import ValidateContext
    from ...RandomGenerator import RandomGenerator
    from ..validation.ValidationFailure import ValidationFailure
    from .VFieldDeclaration import VFieldDeclaration
    from .VTypeDeclaration import VTypeDeclaration
    from ..generation.GenerateContext import GenerateContext

from .VType import VType

_STRUCT_NAME: str = "Object"


class VStruct(VType):

    def __init__(self, name: str, fields: dict[str, 'VFieldDeclaration']) -> None:
        self.name = name
        self.fields = fields

    def get_type_parameter_count(self) -> int:
        return 0

    def validate(self, value: object,
                 type_parameters: list['VTypeDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
        from ..validation.ValidateStruct import validate_struct
        return validate_struct(value, self.name, self.fields, ctx)

    def generate_random_value(self, blueprint_value: object, use_blueprint_value: bool, type_parameters: list['VTypeDeclaration'], ctx: 'GenerateContext') -> object:
        from ..generation.GenerateRandomStruct import generate_random_struct
        return generate_random_struct(blueprint_value, use_blueprint_value, self.fields, ctx)

    def get_name(self) -> str:
        return _STRUCT_NAME

File: lib/java/src/main/java/io/github/telepact/internal/types/VStruct.java
package io.github.telepact.internal.types;

import static io.github.telepact.internal.generation.GenerateRandomStruct.generateRandomStruct;
import static io.github.telepact.internal.validation.ValidateStruct.validateStruct;

import java.util.List;
import java.util.Map;

import io.github.telepact.internal.generation.GenerateContext;
import io.github.telepact.internal.validation.ValidateContext;
import io.github.telepact.internal.validation.ValidationFailure;

public class VStruct implements VType {

    public static final String _STRUCT_NAME = "Object";

    public final String name;
    public final Map<String, VFieldDeclaration> fields;

    public VStruct(String name, Map<String, VFieldDeclaration> fields) {
        this.name = name;
        this.fields = fields;
    }

    @Override
    public int getTypeParameterCount() {
        return 0;
    }

    @Override
    public List<ValidationFailure> validate(Object value, List<VTypeDeclaration> typeParameters, ValidateContext ctx) {
        return validateStruct(value, this.name, this.fields, ctx);
    }

    @Override
    public Object generateRandomValue(Object blueprintValue, boolean useBlueprintValue,
            List<VTypeDeclaration> typeParameters, GenerateContext ctx) {
        return generateRandomStruct(blueprintValue, useBlueprintValue, this.fields, ctx);
    }

    @Override
    public String getName() {
        return _STRUCT_NAME;
    }
}
